build/documentation/generate-typedocs.js

```diff
@@ -0,0 +1,198 @@
+
+    glob(packages, {realpath:true}, function(error, srcFiles) {
+       
+        srcFiles.forEach((srcPath) => {    
+            
+            var valid = true;
+
+            var packageName = srcPath
+                .split(path.sep)
+                .pop();
+
+            excludedPackages.forEach((excludedPackageName) => {
+                if (packageName == excludedPackageName) {
+                    console.log('\x1b[33m%s\x1b[0m', `...skip generating Typedoc API files for ${packageName}`);
+                    valid = false;
+                }
+            });
+
+            if(valid) {
+                dryRun ? console.log(`...generate Typedoc API files for ${packageName}`) : createAPIFiles(packageName, srcPath);
+                dryRun ? console.log(`...add API link to ${packageName}`) : addAPILinkToReadme(packageName);
+            }
+
+        });
+
+    });
+}
+
+/**
+ * Uses Typedoc process to generate API docs for a given package
+ * Stores them in the website/static folder 
+ */
+function createAPIFiles(packageName, srcPath) {
+    
+    var config = path.join(srcPath, '/tsconfig.json');
+    var output = path.join(destDir, packageName, 'api');
+    var file = path.join(srcPath, '/src/index.ts');
+
+    var typedocCmd = os.platform() == 'win32' ? 'typedoc.cmd' : 'typedoc';
+    var test = path.join('node_modules', '.bin', typedocCmd);
+
+    const typedoc = spawn(test, // the local typedoc installation is needed for the markdown theme to work
+        [
+            '--tsconfig', config,
+            '--out', output, file,
+            '--theme', "markdown"
+        ]);
+
+    typedoc.on('close', code => {
+        if (code) {
+            console.log('\x1b[31m%s\x1b[0m', `${packageName} - typedoc API docs not generated, error code: ${code}`);
+        } else {
+            console.log(`${packageName} - typedoc API docs generated`)
+            addHeaderToReadme(packageName);
+        }
+    });
+
+    // sometimes important errors will get sent to stdout for this process
+    // leaving this here because it's useful for debugging, but generates a lot
+    // of output, makes the terminal output harder to read so it's commented out
+
+    // typedoc.stdout.on('data', (data) => {
+    //     console.error(`typedoc stdout:\n${data}`);
+    // });
+
+    typedoc.on('error', (err) => {
+        console.log('\x1b[31m%s\x1b[0m', err);
+    });
+
+}
+
+/**
+ * If the typedoc readme doesn't have this header
+ * It won't be accessible in docusaurus
+ */
+function addHeaderToReadme(packageName) {
+
+    const readmePath = path.join(destDir, packageName, 'api', 'README.md');
+    const readmeText = fs.readFileSync(readmePath).toString();
+
+    var docusaurusHeader = 
+    `---\n` + 
+    `id: index\n` +
+    `---\n\n`;
+
+    try {
+        fs.writeFileSync(readmePath, docusaurusHeader);
+        fs.appendFileSync(readmePath, readmeText);
+    } catch (err) {
+        console.log('\x1b[31m%s\x1b[0m', err);
```

Comment:
maybe add a small comment on what the different regular expressions are doing?

Reply:
Yah I get what you mean, it's close to impossible to see what's happening there. I was thinking of maybe using [chalk](https://github.com/chalk/chalk), since it would add color to the terminal output in a much easier to read manner, but I'm not sure what people would think about me adding another package for something fairly small. Does it matter, or should I just add a small comment describing what's happening?

Reply:
I'm personally ok with taking this small dependency at the root of the project for sake of better logging.

Reply:
Okay, gonna make use of it in that case.